bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java

```diff
@@ -298,13 +300,30 @@ public ResultScanner getScanner(final Scan scan) throws IOException {
     LOG.trace("getScanner(Scan)");
     Span span = TRACER.spanBuilder("BigtableTable.scan").startSpan();
     try (Scope scope = TRACER.withSpan(span)) {
-      clientWrapper.setCaching(scan.getCaching());
-      final ResultScanner scanner = clientWrapper.readRows(hbaseAdapter.adapt(scan));
-      if (hasWhileMatchFilter(scan.getFilter())) {
-        return Adapters.BIGTABLE_WHILE_MATCH_RESULT_RESULT_SCAN_ADAPTER.adapt(scanner, span);
+      if (scan.getCaching() == -1) {
+        final ResultScanner scanner = clientWrapper.readRows(hbaseAdapter.adapt(scan));
+        if (hasWhileMatchFilter(scan.getFilter())) {
+          return Adapters.BIGTABLE_WHILE_MATCH_RESULT_RESULT_SCAN_ADAPTER.adapt(scanner, span);
+        }
+        // TODO: need to end the span when stream ends
+        return scanner;
+      } else {
+        final Query request = hbaseAdapter.adapt(scan);
+        final RequestContext requestContext =
+            RequestContext.create("ProjectId", "InstanceId", "AppProfile");
+        int requestedPageSize = (int) request.toProto(requestContext).getRowsLimit();
+        if (requestedPageSize > scan.getCaching()) {
+          requestedPageSize = scan.getCaching();
+        }
+
+        Query.QueryPaginator paginator = request.createPaginator(requestedPageSize);
+        final ResultScanner scanner = clientWrapper.readRows(paginator, requestedPageSize);
+        if (hasWhileMatchFilter(scan.getFilter())) {
```

Comment:
can we extract this logic out?
ResultScanner resultScanner;
if (getCaching == -1) {
  resultScanner = clientWrapper.readRows(...);
} else {
   Query request = adapter.adapt(...);
   ...
   resultScanner = clientWrapper.readRows(..);
}

if (hasWhileMatchFilter(..)) {
 ..
}
return scanner;

Reply:
Done